core/src/main/java/com/datastax/oss/driver/internal/core/metadata/DefaultTopologyMonitor.java

```diff
@@ -259,69 +251,135 @@ private String retrievePeerTableName() {
     return "system.peers";
   }
 
-  private NodeInfo asNodeInfo(AdminRow row) {
-    return nodeInfoBuilder(row, getBroadcastRpcAddress(row)).build();
-  }
-
-  private Optional<NodeInfo> firstRowAsNodeInfo(AdminResult result) {
+  private Optional<NodeInfo> firstPeerRowAsNodeInfo(AdminResult result) {
     Iterator<AdminRow> iterator = result.iterator();
     if (iterator.hasNext()) {
-      return Optional.of(asNodeInfo(iterator.next()));
+      return buildNodeInfo(iterator.next());
     } else {
       return Optional.empty();
     }
   }
 
   /**
-   * @param broadcastRpcAddress this is a parameter only because we already have it when we come
-   *     from {@link #findInPeers(AdminResult, InetSocketAddress)}. Callers that don't already have
-   *     it can use {@link #getBroadcastRpcAddress}.
+   * Creates a {@link NodeInfo} instance from the given row; returns empty if the row does not
+   * contain enough information, in which case the node will be discarded and won't show up in the
+   * nodes list.
    */
-  protected DefaultNodeInfo.Builder nodeInfoBuilder(
-      AdminRow row, InetSocketAddress broadcastRpcAddress) {
+  @NonNull
+  protected Optional<NodeInfo> buildNodeInfo(@NonNull AdminRow row) {
 
-    // Deployments that use a custom EndPoint implementation will need their own TopologyMonitor.
-    // One simple approach is to extend this class and override this method.
-    EndPoint endPoint =
-        new DefaultEndPoint(context.getAddressTranslator().translate(broadcastRpcAddress));
+    boolean peer = row.contains("peer");
 
-    DefaultNodeInfo.Builder builder =
-        DefaultNodeInfo.builder()
-            .withEndPoint(endPoint)
-            .withBroadcastRpcAddress(broadcastRpcAddress);
-    InetAddress broadcastAddress = row.getInetAddress("broadcast_address"); // in system.local
-    if (broadcastAddress == null) {
-      broadcastAddress = row.getInetAddress("peer"); // in system.peers
+    InetSocketAddress broadcastRpcAddress = getBroadcastRpcAddress(row).orElse(null);
+
+    EndPoint endPoint;
+    if (peer) {
+      // If this node is a peer, it must be valid and its broadcast RPC address must be present,
+      // since we use that address to create the node's endpoint.
+      if (!isPeerValid(row) || broadcastRpcAddress == null) {
+        return Optional.empty();
+      }
+      // Deployments that use a custom EndPoint implementation will need their own TopologyMonitor.
+      // One simple approach is to extend this class and override this method.
+      endPoint = new DefaultEndPoint(context.getAddressTranslator().translate(broadcastRpcAddress));
+    } else {
+      // Don't rely on system.local.rpc_address for the control node, because it mistakenly
+      // reports the normal RPC address instead of the broadcast one (CASSANDRA-11181). We
+      // already know the endpoint anyway since we've just used it to query.
+      endPoint = controlConnection.channel().getEndPoint();
```

Comment:
This hopefully eliminates the need of that awkward downcast in the previous version.

Reply:
One thing I'll note is that `broadcastRpcAddress` is not derived from the endpoint anymore, so it might be wrong (CASSANDRA-11181). But we don't use it to connect anymore so it won't hurt the driver. One might view it as an accurate reflection of the system tables, even if they are wrong.

Reply:
That was precisely my intent.

Reply:
OK, works for me.

Reply:
Actually, I just realized that re-reading the control connection's channel is racy: it could have reconnected since we executed the requests. You can handle that by reading the control endpoint in `refreshNodeList` and passing it into this method.

Reply:
Good catch, will do.